pkg/reportingplugins/mercury/v1/mercury.go

```diff
@@ -0,0 +1,428 @@
+	logger             logger.Logger
+	onchainConfigCodec mercury.OnchainConfigCodec
+	reportCodec        ReportCodec
+}
+
+func NewFactory(ds DataSource, lggr logger.Logger, occ mercury.OnchainConfigCodec, rc ReportCodec) Factory {
+	return Factory{ds, lggr, occ, rc}
+}
+
+func (fac Factory) NewMercuryPlugin(configuration ocr3types.MercuryPluginConfig) (ocr3types.MercuryPlugin, ocr3types.MercuryPluginInfo, error) {
+	offchainConfig, err := mercury.DecodeOffchainConfig(configuration.OffchainConfig)
+	if err != nil {
+		return nil, ocr3types.MercuryPluginInfo{}, err
+	}
+
+	onchainConfig, err := fac.onchainConfigCodec.Decode(configuration.OnchainConfig)
+	if err != nil {
+		return nil, ocr3types.MercuryPluginInfo{}, err
+	}
+
+	maxReportLength, err := fac.reportCodec.MaxReportLength(configuration.N)
+	if err != nil {
+		return nil, ocr3types.MercuryPluginInfo{}, err
+	}
+
+	r := &reportingPlugin{
+		offchainConfig,
+		onchainConfig,
+		fac.dataSource,
+		fac.logger,
+		fac.reportCodec,
+		configuration.ConfigDigest,
+		configuration.F,
+		mercury.EpochRound{},
+		new(big.Int),
+		maxReportLength,
+	}
+
+	return r, ocr3types.MercuryPluginInfo{
+		Name: "Mercury",
+		Limits: ocr3types.MercuryPluginLimits{
+			MaxObservationLength: maxObservationLength,
+			MaxReportLength:      maxReportLength,
+		},
+	}, nil
+}
+
+var _ ocr3types.MercuryPlugin = (*reportingPlugin)(nil)
+
+type reportingPlugin struct {
+	offchainConfig mercury.OffchainConfig
+	onchainConfig  mercury.OnchainConfig
+	dataSource     DataSource
+	logger         logger.Logger
+	reportCodec    ReportCodec
+
+	configDigest             ocrtypes.ConfigDigest
+	f                        int
+	latestAcceptedEpochRound mercury.EpochRound
+	latestAcceptedMedian     *big.Int
+	maxReportLength          int
+}
+
+func (rp *reportingPlugin) Observation(ctx context.Context, repts ocrtypes.ReportTimestamp, previousReport ocrtypes.Report) (ocrtypes.Observation, error) {
+	obs, err := rp.dataSource.Observe(ctx, repts, previousReport == nil)
+	if err != nil {
+		return nil, pkgerrors.Errorf("DataSource.Observe returned an error: %s", err)
+	}
+
+	observationTimestamp := time.Now()
+	if observationTimestamp.Unix() > math.MaxUint32 {
+		return nil, fmt.Errorf("current unix epoch %d exceeds max uint32", observationTimestamp.Unix())
+	}
+	p := MercuryObservationProto{Timestamp: uint32(observationTimestamp.Unix())}
+	var obsErrors []error
+
+	var bpErr, bidErr, askErr error
+	if obs.BenchmarkPrice.Err != nil {
+		bpErr = pkgerrors.Wrap(obs.BenchmarkPrice.Err, "failed to observe BenchmarkPrice")
+		obsErrors = append(obsErrors, bpErr)
+	} else if benchmarkPrice, err := mercury.EncodeValueInt192(obs.BenchmarkPrice.Val); err != nil {
+		bpErr = pkgerrors.Wrap(err, "failed to observe BenchmarkPrice; encoding failed")
+		obsErrors = append(obsErrors, bpErr)
+	} else {
+		p.BenchmarkPrice = benchmarkPrice
+	}
+
+	if obs.Bid.Err != nil {
+		// TODO: Add tests that its invalid if encoding fails on all
```

Comment:
Did I add these? May need to be copied from v0

Reply:
It's all tested in each specific `mercury_test.go`